### PR TITLE
[16.0][ADD] edi_exchange_deduplicate_oca

### PR DESCRIPTION
**Pull request by QuocDuong1306**
This module adds options for deduplication records before sending step on type:
- deduplicate_on_send: check if a fresher one does not exist for the same record. If so, mark the oldest one as obsolete.
- delete_obsolete_records: Delete records marked as obsolete.

### CODE REVIEW

**Review by cyrilmanuel — APPROVED** [review]

File: edi_exchange_deduplicate_oca/models/edi_backend.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        and delete their obsolete records if any.
+        """
+        obsolete_records = self.exchange_record_model.search(
+            self._obsolete_records_domain(record_ids=record_ids)
+        )
+        _logger.info(
+            "EDI Exchange delete records: found %d obsolete records to delete.",
+            len(obsolete_records),
+        )
+        obsolete_records.unlink()
```
Comment (SilvioC2C) [inline edi_exchange_deduplicate_oca/models/edi_backend.py]:
```suggestion
        if obsolete_records:
            obsolete_records.unlink()
```

File: edi_exchange_deduplicate_oca/models/edi_exchange_record.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    @api.model_create_multi
+    def create(self, vals_list):
+        records = super().create(vals_list)
+        for rec in records:
+            check_obsoleted_record = (
+                rec.type_id.direction == "output" and rec.type_id.deduplicate_on_send
+            )
+            if check_obsoleted_record:
+                obsoleted_records = rec.get_older_duplicates()
+                obsoleted_records.write({"edi_exchange_state": "obsolete"})
```
Comment (SilvioC2C) [inline edi_exchange_deduplicate_oca/models/edi_exchange_record.py]:
```suggestion
                if obsoleted_records:
                    obsoleted_records.edi_exchange_state = "obsolete"
```

**Review by SilvioC2C — APPROVED** [review]
LGTM, a few minor changes to do

File: edi_exchange_deduplicate_oca/models/edi_exchange_record.py (hunk 2 above)
Comment (QuocDuong1306) [reply edi_exchange_deduplicate_oca/models/edi_exchange_record.py]: Thank @SilvioC2C , I updated the PR as you suggested

File: edi_exchange_deduplicate_oca/models/edi_backend.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+            len(obsolete_records),
+        )
+        if obsolete_records:
+            obsolete_records.unlink()
+
+    def _obsolete_records_domain(self, record_ids=None):
+        """Domain for obsolete records need to delete."""
+        domain = [
+            ("backend_id", "=", self.id),
+            ("type_id.delete_obsolete_records", "=", True),
```
Comment (simahawk) [inline edi_exchange_deduplicate_oca/models/edi_backend.py]:
```suggestion
```
if the record is obsolete, delete it even if the type's flag has been disabled

File: edi_exchange_deduplicate_oca/models/edi_exchange_record.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+            check_obsoleted_record = (
+                rec.type_id.direction == "output" and rec.type_id.deduplicate_on_send
+            )
+            if check_obsoleted_record:
+                obsoleted_records = rec.get_older_duplicates()
+                if obsoleted_records:
+                    obsoleted_records.edi_exchange_state = "obsolete"
+        return records
+
+    def get_older_duplicates(self, count=None):
```
Comment (simahawk) [inline edi_exchange_deduplicate_oca/models/edi_exchange_record.py]:
```suggestion
    def _edi_get_duplicates(self, count=None):
```

**Review by simahawk — CHANGES_REQUESTED** [review]
Can you please add a USAGE and CONFIGURE section in the readme?

Also explain what's the usage of `block_obsolescence`.

**Review by santostelmo — APPROVED** [review]

File: edi_exchange_deduplicate_oca/models/edi_exchange_record.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+from odoo import api, fields, models
+
+
+class EDIExchangeRecord(models.Model):
+
+    _inherit = "edi.exchange.record"
+
+    edi_exchange_state = fields.Selection(
+        selection_add=[
+            ("obsolete", "Obsolete"),
```
Comment (simahawk) [inline edi_exchange_deduplicate_oca/models/edi_exchange_record.py:13]: we miss the on_delete param I think

**Review by simahawk — APPROVED** [review]

File: edi_exchange_deduplicate_oca/models/edi_exchange_record.py (hunk 5 above)
Comment (QuocDuong1306) [reply edi_exchange_deduplicate_oca/models/edi_exchange_record.py:13]: Updated, thank @simahawk

**Pull request merged**